hazelcast/src/main/java/com/hazelcast/internal/networking/ChannelOutboundHandler.java

```diff
@@ -18,42 +18,105 @@
+ * @param <D> the type of the destination. E.g. a ByteBuffer or a {@link com.hazelcast.util.function.Consumer}.
  * @see EventLoopGroup
  * @see ChannelInboundHandler
  * @see ChannelInitializer
  * @see ChannelErrorHandler
  * @see Channel
  */
-public interface ChannelOutboundHandler<F extends OutboundFrame> {
+public abstract class ChannelOutboundHandler<S, D> extends ChannelHandler<ChannelOutboundHandler, S, D> {
+
+    /**
+     * A callback to indicate that there is something that needs to be written.
+     *
+     * A ChannelOutboundHandler should be able to deal with a spurious wakeup.
+     * So it could be for example there is no
+     * frame for it to write to.
+     *
+     * @return true if the content is fully written and this handler is clean.
+     * @throws Exception if something fails while writing to ByteBuffer. When
+     *                   an exception is thrown, the {@link ChannelErrorHandler} is called.
+     */
+    public abstract HandlerStatus onWrite() throws Exception;
+
+    /**
+     * Initializes the dst ByteBuffer with the value for {@link ChannelOption#SO_SNDBUF}.
+     *
+     * The buffer created is reading mode.
+     */
+    protected final void initDstBuffer() {
+        initDstBuffer(channel.config().getOption(SO_SNDBUF));
+    }
+
+    /**
+     * Initializes the dst ByteBuffer with the configured size.
+     *
+     * The buffer created is reading mode.
+     *
+     * @param sizeBytes the size of the dst ByteBuffer.
+     */
+    protected final void initDstBuffer(int sizeBytes) {
+        initDstBuffer(sizeBytes, null);
+    }
 
     /**
-     * A callback to indicate that the Frame should be written to the destination ByteBuffer.
+     * Initializes the dst ByteBuffer with the configured size.
      *
-     * It could be that a Frame is too big to fit into the ByteBuffer in 1 go; in that case this call will be made
-     * for the same Frame multiple times until write returns true.
+     * The buffer created is reading mode.
      *
-     * @param frame the Frame to write
-     * @param dst   the destination ByteBuffer
-     * @return true if the Frame is completely written
-     * @throws Exception if something fails while writing to ByteBuffer. When an exception is thrown, the
-     *                   {@link ChannelErrorHandler} is called.
+     * @param sizeBytes the size of the dst ByteBuffer.
+     * @param bytes     the bytes added to the buffer. Can be null if nothing
+     *                  should be added.
      */
-    boolean onWrite(F frame, ByteBuffer dst) throws Exception;
+    protected final void initDstBuffer(int sizeBytes, byte[] bytes) {
+        ChannelConfig config = channel.config();
+        ByteBuffer buffer = newByteBuffer(sizeBytes, config.getOption(DIRECT_BUF));
+        if (bytes != null) {
+            buffer.put(bytes);
+        }
+        buffer.flip();
+        dst = (D) buffer;
```

Comment:
If destination is a `ByteBuffer `, what's the need for type `D`?

Reply:
The dst type of the ChannelOuboundHandler is D.. not bytebuffer. This method should only be made if the dst is a bytebuffer; hence the cast.